src/internal/group_api/mod.rs

```diff
@@ -297,47 +359,75 @@ pub(crate) fn get_group_keys<'a>(
-                    Some(
-                        recrypt
-                            .generate_transform_key(
-                                &group_priv_key.into(),
-                                &user_master_pub_key.into(),
-                                &auth.signing_private_key().into(),
-                            )?
-                            .into(),
-                    )
-                } else {
-                    None
-                };
-                (group_pub_key, encrypted_group_key, transform_key)
-            })
+) -> impl Future<Item = GroupCreateResult, Error = IronOxideErr> + 'a {
+    user_api::user_key_list(auth, members)
+        .and_then(move |member_ids_and_keys| {
+            // this will occur when one of the UserIds in the members list cannot be found
+            if member_ids_and_keys.len() != members.len() {
+                // figure out which user ids could not be found in the database and include them in the error message.
+                use std::{collections::HashSet, iter::FromIterator};
+                let desired_members_set: HashSet<&UserId> = HashSet::from_iter(members);
+                let found_members: Vec<UserId> =
+                    member_ids_and_keys.into_iter().map(|(x, _)| x).collect();
+                let found_members_set: HashSet<&UserId> = HashSet::from_iter(&found_members);
+                let diff: HashSet<&&UserId> =
+                    desired_members_set.difference(&found_members_set).collect();
+                futures::future::err(IronOxideErr::UserDoesNotExist(format!(
+                    "Failed to find the following users from the `members` list: {:?}",
+                    diff
+                )))
+            } else {
+                transform::gen_group_keys(recrypt)
+                    .and_then(move |(plaintext, group_priv_key, group_pub_key)| {
+                        // encrypt the group secret to the current user as they will be the group admin
+                        let encrypted_group_key = recrypt.encrypt(
+                            &plaintext,
+                            &user_master_pub_key.into(),
+                            &auth.signing_private_key().into(),
+                        )?;
+                        // Map from UserId to (PublicKey, Optional TransformKey)
+                        // TransformKey is optional because we need it for members, but won't need it
+                        // when we expand this to include admins
+                        let member_info_result: Result<
+                            HashMap<UserId, (PublicKey, Option<TransformKey>)>,
+                            _,
+                        > = member_ids_and_keys
+                            .into_iter()
+                            .map(|(id, user_pub_key)| {
+                                let maybe_transform_key = recrypt.generate_transform_key(
+                                    &group_priv_key.clone().into(),
```

Comment:
is this clone needed? I don't see this being used below...

Reply:
Removing it gives
```cannot move out of `group_priv_key`, a captured variable in an `FnMut` closure```
```move occurs because `group_priv_key` has type `recrypt::api::PrivateKey`, which does not implement the `Copy` trait rustc(E0507)```